src/world/map.c

```diff
@@ -1709,10 +1970,258 @@ static money32 map_set_land_height(int flags, int x, int y, int height, int styl
+	if(ebx&GAME_COMMAND_FLAG_APPLY)
+	{
+		footpath_remove_litter(eax, ebx, map_element_height(eax, ecx));
+		//Hmmm... I have a bug in here where raising the corner of a tile doesn't remove walls on it.
+		map_remove_walls_at(eax, ecx, edx*8-16, edx*8+32);
+	}
+	
+	RCT2_GLOBAL(0x9E2E18, uint32) += 0xC8;
+	
+	//RCT2_CALLFUNC_X(0x663A3A, &eax, &ebx, &ecx, (int *)&edx, (int *)&esi, &edi, &ebp); return ebx;
+	
+	int saved_eax = eax;
+	int saved_ecx = ecx;
+	unsigned int saved_edx = edx;
+	
+	uint8 *pdl = (uint8 *)&edx;
+	uint8 *pdh = (uint8 *)&edx+1;
+	uint8 *pbl = (uint8 *)&ebx;
+	uint8 *pbh = (uint8 *)&ebx+1;
+	
+	/*
+	esi = ecx&0xFFFF; //movzx esi, cx
+	esi <<= 8;
+	esi |= eax;
+	esi >>= 3;
+	esi = RCT2_ADDRESS(RCT2_ADDRESS_TILE_MAP_ELEMENT_POINTERS, uint32)[(int)esi/4];
+	*/
+	esi = (int)map_get_first_element_at(eax/32, ecx/32);
+	
+	*pdh = *pdl;
+	*pdh += 4;
+	
+loc_663A54:
+	eax = map_element_get_type((rct_map_element *)esi);
+	if(eax!=MAP_ELEMENT_TYPE_SCENERY || *pdl>((rct_map_element *)esi)->clearance_height || *pdh<((rct_map_element *)esi)->base_height)
+		goto loc_663AA3;
+	eax = ((rct_map_element *)esi)->properties.scenery.type;
+	eax = RCT2_ADDRESS(RCT2_ADDRESS_SMALL_SCENERY_ENTRIES, uint32)[eax];
+	if(((rct_scenery_entry *)eax)->small_scenery.height>0x40) //cmp byte ptr [eax+0Ah], 40h; jbe short loc_663A87
+	{
+		if(RCT2_GLOBAL(RCT2_ADDRESS_PARK_FLAGS, uint32)&PARK_FLAGS_FORBID_TREE_REMOVAL)
+		{
+			//rct2: 0x663C45
+			map_obstruction_set_error_text((rct_map_element *)esi);
+			return MONEY32_UNDEFINED;
+		}
+	}
+	//rct2: 0x663A87
+	//callcode_push3(0x663A87, saved_eax, saved_ecx, saved_edx, &eax, &ebx, &ecx, (int *)&edx, (int *)&esi, &edi, &ebp); return ebx;
+	ecx = ((rct_scenery_entry *)eax)->small_scenery.removal_price*10;
+	RCT2_GLOBAL(0x9E2E18, uint32) += ecx;
+	//rct2: 0x663A94
+	if((ebx&GAME_COMMAND_FLAG_APPLY))
+	{
+		map_element_remove((rct_map_element *)esi);
+		esi = esi-8;
+	}
+loc_663AA3:
+	esi = esi+8;
+	if(!((*(uint8 *)esi-7)&MAP_ELEMENT_FLAG_LAST_TILE))
+		goto loc_663A54;
+	edx = saved_edx;
+	ecx = saved_ecx;
+	eax = saved_eax;
+	saved_eax = eax;
+	int saved_ebx = ebx;
+	saved_ecx = ecx;
+	
+	//callcode_push3(0x663AB2, saved_eax, saved_ebx, saved_ecx, &eax, &ebx, &ecx, (int *)&edx, (int *)&esi, &edi, &ebp); return ebx;
+	
+	/*
+	esi = ecx&0xFFFF; //movzx esi, cx
+	esi <<= 8;
+	esi |= eax&0xFFFF; //or si, ax
+	esi >>= 3;
+	esi = RCT2_ADDRESS(RCT2_ADDRESS_TILE_MAP_ELEMENT_POINTERS, uint32)[esi/4];
+	*/
+	rct_map_element *mapElement = map_get_first_element_at(eax/32, ecx/32); //esi
+	esi = (int)mapElement;
+	//callcode_push3(0x663AC4, saved_eax, saved_ebx, saved_ecx, &eax, &ebx, &ecx, (int *)&edx, (int *)&esi, &edi, &ebp); return ebx;
+	
+loc_663AC4:
+	eax = mapElement->type;
+	eax &= 0x3C;
+	if(eax==8)
+	{
+		ebp = ((rct_map_element *)mapElement)->properties.track.ride_index; //not sure what union member we need here. [esi+7]
+		eax = RCT2_ADDRESS(RCT2_ADDRESS_RIDE_LIST, rct_ride)[ebp].subtype;
```

Comment:
Use `rct_ride* ride = GET_RIDE(mapElement->properties.track.ride_index);` followed by
`rct_ride_type* rideEntry = GET_ENTRY(ride->subtype);` for lines 2077->2079.